Array assignments with --public-flat-rw

### Verilog file
```
module arr_test ();

  logic din [0:15];

  array_test array_test_inst(.din(din));

endmodule


module array_test(
    input din [0:15]
);

endmodule
```
### Error

```
make[1]: Entering directory '/home/nndurj/Work/scratch/cocotb/examples/adder/tests'
verilator -cc --exe -Mdir sim_build -DCOCOTB_SIM=1 --top-module arr_test --vpi --public-flat-rw --prefix Vtop -o arr_test -LDFLAGS "-L `dirname /home/nndurj/Work/scratch/cocotb/examples/adder/tests/build/libs/x86_64/libvpi.so` -lvpi -lgpi -lcocotb -lgpilog -lcocotbutils"  /home/nndurj/Work/scratch/cocotb/examples/adder/tests/../hdl/adder.v /home/nndurj/miniconda3/envs/cocotb_verilator/lib/python3.7/site-packages/cocotb/share/lib/verilator/verilator.cpp
CPPFLAGS="-DVL_TIME_PRECISION_STR=1ps -std=c++11" make -C sim_build -f Vtop.mk
make[2]: Entering directory '/home/nndurj/Work/scratch/cocotb/examples/adder/tests/sim_build'
g++  -DVL_TIME_PRECISION_STR=1ps -std=c++11 -I.  -MMD -I/home/nndurj/softwares/verilator/include -I/home/nndurj/softwares/verilator/include/vltstd -DVM_COVERAGE=0 -DVM_SC=0 -DVM_TRACE=0 -Wno-sign-compare -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-unused-variable -Wno-shadow       -c -o verilator.o /home/nndurj/miniconda3/envs/cocotb_verilator/lib/python3.7/site-packages/cocotb/share/lib/verilator/verilator.cpp
/usr/bin/perl /home/nndurj/softwares/verilator/bin/verilator_includer -DVL_INCLUDE_OPT=include Vtop.cpp > Vtop__ALLcls.cpp
g++  -DVL_TIME_PRECISION_STR=1ps -std=c++11 -I.  -MMD -I/home/nndurj/softwares/verilator/include -I/home/nndurj/softwares/verilator/include/vltstd -DVM_COVERAGE=0 -DVM_SC=0 -DVM_TRACE=0 -Wno-sign-compare -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-unused-variable -Wno-shadow       -c -o Vtop__ALLcls.o Vtop__ALLcls.cpp
In file included from Vtop__ALLcls.cpp:3:0:
Vtop.cpp: In static member function ‘static void Vtop::_settle__TOP__1(Vtop__Syms*)’:
Vtop.cpp:135:9: error: invalid array assignment
         = vlTOPp->arr_test__DOT____Vcellinp__array_test_inst__din;
         ^
Vtop.cpp:137:9: error: invalid array assignment
         = vlTOPp->arr_test__DOT____Vcellinp__array_test_inst__din;
         ^
/home/nndurj/softwares/verilator/include/verilated.mk:182: recipe for target 'Vtop__ALLcls.o' failed
make[2]: *** [Vtop__ALLcls.o] Error 1
make[2]: Leaving directory '/home/nndurj/Work/scratch/cocotb/examples/adder/tests/sim_build'
/home/nndurj/miniconda3/envs/cocotb_verilator/lib/python3.7/site-packages/cocotb/share/makefiles/simulators/Makefile.verilator:44: recipe for target 'sim_build/arr_test' failed
make[1]: *** [sim_build/arr_test] Error 2
make[1]: Leaving directory '/home/nndurj/Work/scratch/cocotb/examples/adder/tests'
/home/nndurj/miniconda3/envs/cocotb_verilator/lib/python3.7/site-packages/cocotb/share/makefiles/Makefile.sim:79: recipe for target 'sim' failed
make: *** [sim] Error 2
```

May we assist you in trying to fix this yourself?
Yes

--- a/src/V3Inline.cpp
+++ b/src/V3Inline.cpp
@@ -322,6 +322,15 @@ private:
                     new AstAssignW(nodep->fileline(),
                                    new AstVarRef(nodep->fileline(), exprvarrefp->varp(), true),
                                    new AstVarRef(nodep->fileline(), nodep, false)));
+            } else if (nodep->isSigPublic() && VN_IS(nodep->dtypep(), UnpackArrayDType)) {
+                // Public variable at this end and it is an unpacked array. We need to assign
+                // instead of aliased, because otherwise it will pass V3Slice and invalid
+                // code will be emitted.
+                UINFO(9,"assign to public and unpacked: "<<nodep<<endl);
+                m_modp->addStmtp(
+                    new AstAssignW(nodep->fileline(),
+                                   new AstVarRef(nodep->fileline(), exprvarrefp->varp(), true),
+                                   new AstVarRef(nodep->fileline(), nodep, false)));
             } else if (nodep->isIfaceRef()) {
                 m_modp->addStmtp(
                     new AstAssignVarScope(nodep->fileline(),

--- a/test_regress/t/t_array_unpacked_public.pl
+++ b/test_regress/t/t_array_unpacked_public.pl
@@ -1,0 +1,21 @@
+#!/usr/bin/perl
+if (!$::Driver) { use FindBin; exec("$FindBin::Bin/bootstrap.pl", @ARGV, $0); die; }
+# DESCRIPTION: Verilator: Verilog Test driver/expect definition
+#
+# Copyright 2003 by Wilson Snyder. This program is free software; you can
+# redistribute it and/or modify it under the terms of either the GNU
+# Lesser General Public License Version 3 or the Perl Artistic License
+# Version 2.0.
+
+scenarios(simulator => 1);
+
+compile(
+    verilator_flags2 => ["--public-flat-rw"],
+    );
+
+execute(
+    check_finished => 1,
+    );
+
+ok(1);
+1;

--- a/test_regress/t/t_array_unpacked_public.v
+++ b/test_regress/t/t_array_unpacked_public.v
@@ -1,0 +1,21 @@
+// DESCRIPTION: Verilator: Verilog Test module
+//
+// This file ONLY is placed into the Public Domain, for any use,
+// without warranty, 2020 by Stefan Wallentowitz
+
+module t();
+  logic din [0:15];
+
+  array_test array_test_inst(.din(din));
+
+  initial begin
+    $write("*-* All Finished *-*\n");
+    $finish;
+  end
+endmodule
+
+module array_test(
+    input din [0:15]
+);
+
+endmodule

